### ISSUE TITLE
Loading of game objects/assets fails with a vanilla extraction of RCT2 assets on case-sensitive filesystems on OSX/MacOS

### ISSUE THREAD

**Issue by ooshlablu** [OpenRCT2/OpenRCT2#8596]
**OS:** OSX/MacOS
**Version:** 10.14.2
**Commit/Build:** 8ac731e

The loading of game assets does not appear to work out of the box on case-sensitive file systems on OSX/MacOS.

**This entire section seems irrelevant for non-OSX/MacOX Operating systems, see comments below:**
From what I can tell from the verbose output from the game, it is looking to load several object types at startup, for instance `*.dat`, `*.pob`, `*.json`, `*.parkobj`, etc (**note**: this applies to _all_ object types, including music, ride sounds, rides, etc). From a brief look at the code (https://github.com/OpenRCT2/OpenRCT2/blob/9f891bfdcc5cceb999fe69137754a7b54a8b2f4b/src/openrct2/FileClassifier.cpp), the game is matching those _exact_ extensions, and is not picking up the files the way they are named when they are extracted from the installation media (in my case: retail CD), which can have extensions such as `*.DAT`, `*.POB`, etc (upper-case). Since NTFS is case-insensitive, these extensions would be easily recognized, and no issue would appear. When trying to run the game with the capitalized extensions on case-sensitive HFS+, starting a game fails with :`The file could not be loaded as some of the objects are missing or corrupted`.

I was able to work around this by entering the directories (both the extracted RCT2 directories and openrct2's `object` directory) and copying to lower-case extensions using this silly bash loop: `IFS=$'\n'; EXTENSION='.dat'; for FILE in $(ls -1 | grep -i ${EXTENSION}); do NAME=$(echo ${FILE} | cut -d. -f1); cp ${FILE} ${NAME}.${EXTENSION}; done`

**Steps to reproduce:**
1. Must have a Mac formatted with a case-sensitive HFS+ filesystem (which I believe is default on 2015+ Models)
2. Follow the Mac installation instructions from openrct2.org
3. Run the game, start a scenario, and it will fail with `The file could not be loaded as some of the objects are missing or corrupted`
4. Exit the game, `cd` into each of the asset directories that have upper-case extensions and run the loop above (**note**: be sure to change `.dat` in the `EXTENSION` variable in the loop above to the correct extension for the assets in that directory)
5. Run the game, start a scenario, and it'll work!!!!

I haven't done a ton of digging yet, and will test on Linux either later tonight or tomorrow and update the ticket with that info as well. I also intend to fork the project and take a look at the code to see if I can parse both types of extensions (upper and lower case), although my C/C++ skills are....... basically non-existent. If I figure it out, a PR will be coming your way!

**Comment by janisozaur** [comment]
Running on case-sensitive filesystems should not be a problem. If it is, that must be platform-specific, as I've been running OpenRCT2 on case-sensitive filesystems on various Linuxes and derivatives without any issues.

Here's how my RCT2 installation looks like, I removed some parts of it, as there is nothing different about them.
```
rct2
├── Data
│   ├── ch.dat
│   ├── css10.dat
│   ├── css11.dat
│   ├── css12.dat
│   ├── css13.dat
│   ├── css14.dat
│   ├── css15.dat
│   ├── css16.dat
│   ├── css17.dat
│   ├── css18.dat
│   ├── css19.dat
│   ├── css1.dat
│   ├── css20.dat
│   ├── css21.dat
│   ├── css22.dat
│   ├── css23.dat
│   ├── css24.dat
│   ├── css25.dat
│   ├── css26.dat
│   ├── css27.dat
│   ├── css28.dat
│   ├── css29.dat
│   ├── css2.dat
│   ├── css30.dat
│   ├── css31.dat
│   ├── css32.dat
│   ├── css33.dat
│   ├── css34.dat
│   ├── css35.dat
│   ├── css36.dat
│   ├── css37.dat
│   ├── css38.dat
│   ├── css39.dat
│   ├── css3.dat
│   ├── css40.dat
│   ├── css41.dat
│   ├── css42.dat
│   ├── css43.dat
│   ├── css44.dat
│   ├── css45.dat
│   ├── css46.dat
│   ├── css4.dat
│   ├── css5.dat
│   ├── css6.dat
│   ├── css7.dat
│   ├── css8.dat
│   ├── css9.dat
│   ├── CUSTOM.TXT
│   ├── g1.dat
│   ├── game.cfg
│   ├── hang.dat
│   ├── kanji.dat
│   ├── tai.dat
│   ├── tut640a.dat
│   ├── tut640b.dat
│   ├── tut640c.dat
│   ├── tut800a.dat
│   ├── tut800b.dat
│   ├── tut800c.dat
│   ├── tutc640a.dat
│   ├── tutc640b.dat
│   ├── tutc640c.dat
│   ├── tutc800a.dat
│   ├── tutc800b.dat
│   ├── tutc800c.dat
│   ├── tutk640a.dat
│   ├── tutk640b.dat
│   ├── tutk640c.dat
│   ├── tutk800a.dat
│   ├── tutk800b.dat
│   └── tutk800c.dat
├── Landscapes
│   ├── 001
│   └── Coasterization of Mars.SC6
├── License.txt
├── MANUAL.PDF
├── ObjData
│   ├── 1920PATH.DAT
│   ├── 1920RACR.DAT
│   ├── 1920SAND.DAT
│   ├── 1920SENT.DAT
│   ├── 1920SLMP.DAT
│   ├── 1950SCAR.DAT
│   ├── 1960TSRT.DAT
│   ├── 1TBN.DAT
│   ├── 1X1ATRE2.DAT
│   ├── 1X1ATREE.DAT
│   ├── 1X1BRANG.DAT
│   ├── 1X1DIDGE.DAT
│   ├── 1X1EMUXX.DAT
│   ├── 1X1JUGT2.DAT
│   ├── 1X1JUGT3.DAT
│   ├── 1X1KANGA.DAT
│   ├── 1X1TERMM.DAT
│   ├── 1X2ABR01.DAT
│   ├── 1X2ABR02.DAT
│   ├── 1X2ABR03.DAT
│   ├── 1X2AZT01.DAT
│   ├── 1X2AZT02.DAT
│   ├── 1X2GLAMA.DAT
│   ├── 1X2LGCHM.DAT
│   ├── 1X2LION.DAT
│   ├── 1X2RHINO.DAT
│   ├── 1X2TRCTR.DAT
│   ├── 1X2ZEBRA.DAT
│   ├── 1X4BRG01.DAT
│   ├── 1X4BRG02.DAT
│   ├── 1X4BRG03.DAT
│   ├── <snip>
│   ├── WTUDOR12.DAT
│   ├── WTUDOR13.DAT
│   ├── WTUDOR14.DAT
│   ├── WTUDOR15.DAT
│   ├── WTUDOR16.DAT
│   ├── WTUDOR17.DAT
│   ├── WTUDOR18.DAT
│   ├── WWBANK.DAT
│   ├── WWIND03.DAT
│   ├── WWIND04.DAT
│   ├── WWIND05.DAT
│   ├── WWIND06.DAT
│   ├── WWTWA.DAT
│   ├── WWTW.DAT
│   ├── YEWTREEX.DAT
│   ├── ZEPLELIN.DAT
│   ├── ZEUSSTAT.DAT
│   ├── ZLDB.DAT
│   └── ZLOG.DAT
├── Readme_rct.txt
├── ReadMe.txt
├── Saved Games
│   ├── em.SV6
│   └── scores.dat
├── Scenarios
│   ├── Africa - African Diamond Mine.SC6
│   ├── Africa - Oasis.SC6
│   ├── Africa - Victoria Falls.SC6
│   ├── Alpine Adventures.SC6
│   ├── Amity Airfield.SC6
│   ├── Antarctic - Ecological Salvage.SC6
│   ├── Asia - Great Wall of China Tourism Enhancement.SC6
│   ├── <snip>
│   ├── Six Flags over Texas.SC6
│   ├── South America - Inca Lost City.SC6
│   ├── South America - Rain Forest Plateau.SC6
│   └── South America - Rio Carnival.SC6
└── Tracks
    ├── Aerial Cycles.TD6
    ├── Atomizer.TD6
    ├── Automania.TD6
    ├── Automobile Antics.TD6
    ├── Bat Flight.TD6
    ├── Batty Bikers.TD6
    ├── Bear Circle.TD6
    ├── <snip>
    ├── Twicky Twister.TD6
    ├── Twin Loop.TD6
    ├── Twisty Towers.TD6
    ├── Ugly Twisters.TD6
    ├── Updraft.TD6
    ├── Velocerator.TD6
    ├── Venomous.TD6
    ├── Verticulator.TD6
    ├── Viper.TD6
    ├── Wheely Reel.TD6
    ├── Whiteout.TD6
    ├── Whitewash.TD6
    ├── Wicked Whizzer.TD6
    ├── Wings of Apollo.TD6
    ├── Woodchip.TD6
    ├── Woodwind.TD6
    ├── Woodworm.TD6
    └── X.TD6

10 directories, 7153 files
```
All of the distributions of the game I know of have the very same format.

**Comment by janisozaur** [comment]
@Gymnasiast do you have retail CD? Mine was probably lost years ago. Can you verify what's the structure there?

**Comment by Gymnasiast** [comment]
@janisozaur The structure and casing are the same on my retail installation. Unlike RCT1, the files and structure of the CD/GOG/Steam versions are very much alike for RCT2.

**Comment by ooshlablu** [comment]
Follow up:
I was able to build it and run it on Ubuntu 18.04 by doing the following:
```
sudo apt-get install --no-install-recommends -y cmake libsdl2-dev libicu-dev gcc pkg-config libjansson-dev libspeex-dev libspeexdsp-dev libcurl4-openssl-dev libcrypto++-dev libfontconfig1-dev libfreetype6-dev libpng-dev libssl-dev libzip-dev git
git clone https://github.com/OpenRCT2/OpenRCT2.git
cd OpenRCT2/build
cmake ../
make
DESTDIR=. make install
## In My case, ~/games/rct2/RCT2/ is where I extracted the game components, no tricks
./openrct2 set-rct2 ~/games/rct2/RCT2/
./openrct2 ## Run the game
```
and there's some parts that work which I didn't even notice on the Mac install, like the roller coaster graphics in the main menu screen, that work straight-out-of-the-box!!!

I didn't have to copy/rename any files, and it just straight-up works.

So now I'm baffled. I have another Mac with HFS+ case-sensitive that I can try on tomorrow, and a mac-mini in my basement I can run Vagrant images on for previous versions of OSX/MacOS. I'll let you know what I find.

**Comment by ooshlablu** [comment]
I updated the title of the issue to reflect that this appears to be a Mac only issue now

**Comment by janisozaur** [comment]
> I'm trying to install on Ubuntu 18.04 using the Quick Start instructions from https://openrct2.org/quickstart/install/linux/debian-ubuntu-mint

If you used our official website you wouldn't run into such problems, it's kept updated much better. Please file your complaint with https://github.com/Limetric/OpenRCT2.org instead.

**Comment by janisozaur** [comment]
You may want to specify the build type with cmake, as explained in our docs, e.g. `-DCMAKE_BUILD_TYPE=RelWithDebInfo`

**Comment by IntelOrca** [comment]
@ooshlablu Our code that matches file extensions is always a case insensitive comparison. If there is a problem with case sensitivity, it is usually because one or more directories is the wrong case. There is some code to try and find directories of the same name but different case but it isn't perfect. Was it only the file names you changed to fix your macOS issue or the directories as well?

**Comment by ooshlablu** [comment]
@IntelOrca It was only the file names, I did not change the directory structure at all

**Comment by ooshlablu** [comment]
@janisozaur I'm confused by your comment, I thought openrct2.org was the official website.

The PPA appears to be dead, so if it isn't useful, it should be removed from the docs to prevent confusion from users who don't have in-depth knowledge. It will turn them off from the project. I will take a look at that repo, thanx!

**Comment by Gymnasiast** [comment]
The PPA is most certainly not dead. The master branch of it is, however, and .ORG still lists old instructions (so did .IO until earlier this month, oops!)

The official website is openrct2.io, but openrct2.org has been serving as the go-to site for most users. Janisozaur likes to pretend we have nothing to do with it, but it's more complicated than that.

```
sudo add-apt-repository ppa:openrct2/nightly
sudo apt-get install openrct2
```

**Comment by ooshlablu** [comment]
@Gymnasiast thanx for the clarification

**Comment by IntelOrca** [comment]
@ooshlablu could you send us the verbose output from the game on your macOS machine with the original file names?

**Comment by tomlankhorst** [comment]
I'm experiencing a comparable problem on `<develop>` on Mac OS 10.14.2. Followed instructions and pointed the game to a GOG install of RCT2 Triple Thrill, installed in some `~/.wine/drive_c` directory. 

Starting the game leaves me with a black intro scenario and then a loading error when I try to load any park. 

`The file could not be loaded as some of the objects are missing or corrupted`

![Screenshot of RCT Mac OS](https://user-images.githubusercontent.com/675432/51372858-d7707e80-1afe-11e9-9906-9b8253fa6de3.png)

Other resource errors pop up too: 

`ERROR[/OpenRCT2/src/openrct2/scenario/ScenarioRepository.cpp:557 (AddScenario)]: Tried to add scenario with an empty filename!`

Is this the same issue? Shall I provide debug logs?

**Comment by IntelOrca** [comment]
@tomlankhorst all your objects should be obtained from JSON files which are distributed with OpenRCT2. When building from source you typically download the latest objects zip (this is automatic via cmake install or the VS/xcode projects) and have that next the built binary.

**Comment by ooshlablu** [comment]
Sorry for the long reply @IntelOrca attached is the verbose log from a fresh install on a different OSX machine. Let me know what you want to do next. I still intend to run the vagrant boxes, but it will be a few days or so.
[openrct2_osx.log](https://github.com/OpenRCT2/OpenRCT2/files/2777502/openrct2_osx.log)

**Comment by IntelOrca** [comment]
@ooshlablu have you tried using the latest development version? You can get a build from https://openrct2.org/downloads.

**Comment by jgreco** [comment]
@tomlankhorst I had this same issue after installing OpenRCT2 via homebrew.  However using OpenRCT2 from [this zip](https://github.com/OpenRCT2/OpenRCT2/releases/download/v0.2.1/OpenRCT2-0.2.1-macos.zip) resolved the issue.

From what I've gathered from @IntelOrca's response, it seems the homebrew formula for openrct2 is probably skipping a few steps.  It also had some permissions issues for me with `/usr/local/share/icons/hicolor/64x64/apps`  It seems to me that the homebrew formula needs some TLC.

**Comment by ooshlablu** [comment]
@IntelOrca The dev build appears to work perfectly fine without having to rename/copy anything, at least on 10.14.2.

Out of curiosity, do you know what the earliest version of OSX the pre-built apps can be run on?

**Comment by ooshlablu** [comment]
Hey guys, I was able to reproduce the issue on OSX 10.13.6 (High Sierra) using the current "release", but was not able to reproduce on earlier versions of OSX all the way down to Yosemite. The dev build works perfectly fine in all cases. I'm going to assume that this will not be an issue in the next release since the dev build appears to work perfectly and will close this issue.

Again, I love what you guys have done with this project, and appreciate your responsiveness in addressing issues. I look forward to many years of childhood memories being instilled in the next generation :-)

You guys rock!

**Comment by janisozaur** [comment]
@ooshlablu while I'm happy the issue seems fixed, I checked the commits since v0.2.1 until now that touched `src/openrct2/platform` and saw nothing I could attribute this change to. I don't recall changing the precompiled libraries either, so could you please try bisecting to find when it started working?
